feat(query-bar): allow to specify hints COMPASS-7829

luca-nardelli opened the pull request:
## Description
This PR adds a new input to the query-bar component to allow to specify the `hint` options of a query, in order to force index selection in the GUI. This is useful when testing different indexes in the "Explain" modal
<img width="1168" alt="image" src="https://github.com/mongodb-js/compass/assets/6215162/bc79e06c-a471-4874-ad2d-f0dcfe98883e">

### Checklist
- [ ] New tests and/or benchmarks are included 
- [ ] Documentation is changed or added
- [x] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it's updating a dependancy, link to the Pull Request that originally introduced the fix -->
- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Open Questions
- Could you point me to relevant tests and/or documnentation changes that would need to happen? 
- In order to limit the scope of this PR, I used the same validation rules as the sort field, is this acceptable or do I need to submit a pr to mongodb-query-parser as well?

## Dependents
<!--- If applicable, link PRs/commits that this PR is dependent on or is a dependency of. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

addaleax commented on packages/compass-query-bar/src/constants/query-option-definition.ts:
```diff
@@ -33,6 +33,12 @@ export const OPTION_DEFINITION: {
     link: 'https://docs.mongodb.com/manual/reference/method/cursor.sort/',
   },
+  hint: {
+    name: 'hint',
```
I'd consider using `Index Hint` or similar for the UI, somebody who isn't familiar with this database feature might not be aware of what kind of `Hint` this is 🙂

luca-nardelli replied:
This makes perfect sense, added it! I noticed though that the `label` field was not being used in https://github.com/mongodb-js/compass/pull/5666/commits/0eae3edfbcc920bb355e24922b9e6a3344054bbd#diff-2d6d3eaf924ff6725b2a090da228b745956458a31e338881ae9d29ab5d712f6bR178 so I added that as well. I believe this also changes how `Max Time MS` is displayed (before it was MaxTimeMS) since it's using a label property as well

![image](https://github.com/mongodb-js/compass/assets/6215162/b8a8a2f4-e34c-4436-8b36-ce07dd8b46f3)

addaleax commented on packages/compass-query-bar/src/utils/query.ts:
```diff
@@ -149,6 +149,11 @@ export function validateField(
     }
   }
 
+  // HACK to validate hint document as a sort document
```
I think that's totally okay and not even really a hack, I'd maybe just explain in this comment that we don't have a validator for indexes so the sort validator works fine as a replacement?

luca-nardelli replied:
Updated!

addaleax commented:
@luca-nardelli Thank you for the PR! 💙

addaleax commented:
@luca-nardelli Looks like there are some failures in the unit tests here:

```
 [2024/04/09 13:57:22.763] /System/Volumes/Data/data/mci/da960ea590c9cfa910cbf4c269753e55/src/packages/compass-query-bar/src/utils/query.ts
 [2024/04/09 13:57:22.763]   3:19  error  'values' is defined but never used  @typescript-eslint/no-unused-vars
 [2024/04/09 13:57:22.763] 
```

```
 [2024/04/09 14:55:42.933] @mongodb-js/compass-query-bar:   91 passing (661ms)
 [2024/04/09 14:55:42.933] @mongodb-js/compass-query-bar:   5 pending
 [2024/04/09 14:55:42.933] @mongodb-js/compass-query-bar:   2 failing
 [2024/04/09 14:55:42.933] @mongodb-js/compass-query-bar:   1) QueryBar Component
 [2024/04/09 14:55:42.933] @mongodb-js/compass-query-bar:        when expanded
 [2024/04/09 14:55:42.933] @mongodb-js/compass-query-bar:          renders the expanded inputs:
 [2024/04/09 14:55:42.933] @mongodb-js/compass-query-bar:       AssertionError: expected 8 to equal 7
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:       + expected - actual
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:       -8
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:       +7
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:       
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:       at Context.<anonymous> (/System/Volumes/Data/data/mci/0257f97a39c461fcc09fd5a5ca1fe03e/src/packages/compass-query-bar/src/components/query-bar.spec.tsx:83:55)
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:       at process.processImmediate (node:internal/timers:476:21)
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar: 
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:   2) queryBarReducer
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:        explainQuery
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:          should call localAppStorage with a correct query:
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:      AssertionError: expected emit to have been called exactly once with arguments 'open-explain-plan-modal', {
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:   query: {
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     filter: { _id: [Object] },
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     collation: null,
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     sort: null,
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     skip: 0,
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     limit: 10,
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     maxTimeMS: 60000,
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     projection: { _id: 1 }
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:   }
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar: }
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar: '"open-explain-plan-modal"'
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar: {
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:   query: {
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     filter: { _id: [Object] },
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     collation: null,
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     sort: null,
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     hint: null,
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     skip: 0,
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     limit: 10,
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     maxTimeMS: 60000,
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:     projection: { _id: 1 }
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:   }
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar: }
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:       at Context.<anonymous> (/System/Volumes/Data/data/mci/0257f97a39c461fcc09fd5a5ca1fe03e/src/packages/compass-query-bar/src/stores/query-bar-reducer.spec.ts:181:68)
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar:       at process.processImmediate (node:internal/timers:476:21)
 [2024/04/09 14:55:42.934] @mongodb-js/compass-query-bar: 
```

luca-nardelli commented:
@addaleax thanks for the message. I updated those 2 tests taking into account the new`hint` option, and removed the unused import. Ran `npm run test-changed` locally and it went through

addaleax commented:
Thank you for the update! There was one more failing test but that seems to already have been resolved in another PR, so I've merged `main` here just to try to get a fully green CI 🙂

addaleax commented:
@luca-nardelli Thank you again for the PR! ✨

luca-nardelli commented:
You're welcome! Looking forward to downloading the new version 😄




Il giorno mar 16 apr 2024 alle 12:30 Anna Henningsen <
***@***.***> ha scritto:

> @luca-nardelli <https://github.com/luca-nardelli> Thank you again for the
> PR! ✨
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mongodb-js/compass/pull/5666#issuecomment-2058763155>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABPNL6Q5UWUWMV3XGV4DT3DY5T4UTAVCNFSM6AAAAABF4V3XSWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANJYG43DGMJVGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
